[Sanitizers] Add Scudo support

Lukasa opened the pull request:
LLVM ships a hardened memory allocator called Scudo:
https://llvm.org/docs/ScudoHardenedAllocator.html. This allocator
provides additional mitigations against heap-based vulnerabilities, but
retains sufficient performance to be safely run in production
applications.

While ideal Swift applications are obviously written in pure Swift, in
practice most applications contain some amount of code written in
less-safe languages. Additionally, plenty of Swift programs themselves
contain unsafe code, particularly when attempting to implement
high-performance data structures. These sources of unsafety introduce
the risk of memory issues, and having the option to use the Scudo
allocator is a useful defense-in-depth tool.

This patch enables `-sanitize=scudo` as an extra `swiftc` flag. This
sanitizer is only supported on Linux, so no further work is required to
enable it on Windows or Apple platforms. As this "sanitizer" is only a
runtime component, we do not require any wider changes to instrument
code.

The result of this patch matches the API of clang, which supports
`-fsanitize=scudo`.

- - - 

I do have one question: currently this patch contains no regression tests. I was fundamentally unsure of where to place these regression tests, but I think we should have them, so please do as part of your review suggest a location and form for the regression tests.

Lukasa committed:
[Sanitizers] Add Scudo support

LLVM ships a hardened memory allocator called Scudo:
https://llvm.org/docs/ScudoHardenedAllocator.html. This allocator
provides additional mitigations against heap-based vulnerabilities, but
retains sufficient performance to be safely run in production
applications.

While ideal Swift applications are obviously written in pure Swift, in
practice most applications contain some amount of code written in
less-safe languages. Additionally, plenty of Swift programs themselves
contain unsafe code, particularly when attempting to implement
high-performance data structures. These sources of unsafety introduce
the risk of memory issues, and having the option to use the Scudo
allocator is a useful defense-in-depth tool.

This patch enables `-sanitize=scudo` as an extra `swiftc` flag. This
sanitizer is only supported on Linux, so no further work is required to
enable it on Windows or Apple platforms. As this "sanitizer" is only a
runtime component, we do not require any wider changes to instrument
code. This is similar to clang's `-fsanitize=scudo` flag.

The Swift driver rejects platforms that don't support Scudo using an
existing mechanism in the Driver that is not part of this patch. This
mechanism is in swift::parseSanitizerArgValues(...)
(lib/Option/SanitizerOptions.cpp). The mechanism determines if a
sanitizer is supported by checking for the existence of the
corresponding sanitizer runtime library in the compiler's resource
directory. The Scudo runtime library currently only exists in the
Linux compiler resource directory. This results in the driver only
allowing Scudo when targeting Linux.

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
I should note that, when we update the version of llvm we're using, enabling Scudo will also enable [GWP-Asan](https://llvm.org/docs/GwpAsan.html), which provides not just mitigation but active detection of problems.

danliew-apple commented:
@Lukasa Thanks for the patch. There are something things that need to be addressed before this can land.

1. Given that this only works on Linux you need to add code in the Swift driver to refuse to use `-sanitize=scudo` on non Linux platforms. I can't see anything obvious in the code you added that forbids this.
2. This patch doesn't change anything about the compilation process. AFAIK Scudo is a replacement allocator library so I would expect that the linker would need to be told to link against the scudo runtime.
3. This is missing tests. You need:

* A test for successful driver invocation on Linux. You should probably check that the linker flags look right (i.e. links against Scudo runtime).
* A test for driver failure on non-Linux platforms.
* Tests for the invalid santizer+scudo combination.
* At least one test case where we compile and run a program that uses Scudo.

Lukasa commented:
Thanks @danliew-apple!

> Given that this only works on Linux you need to add code in the Swift driver to refuse to use -sanitize=scudo on non Linux platforms. I can't see anything obvious in the code you added that forbids this.

I don't believe this is true. There are separate Sanitizer whitelists for [Darwin](https://github.com/apple/swift/blob/3ff6c1c315677789cc96a84c9e17ab9f5105614d/lib/Driver/DarwinToolChains.cpp#L304) and [Windows](https://github.com/apple/swift/blob/238d097af906ea59a7343142d5b28862b2bfc97a/lib/Driver/WindowsToolChains.cpp#L162) which explicitly whitelist the sanitisers they support. I have observed that with this toolchain built on Darwin the Scudo flag is rejected: I believe this is true for Windows as well, though @compnerd could further clarify.

> This patch doesn't change anything about the compilation process. AFAIK Scudo is a replacement allocator library so I would expect that the linker would need to be told to link against the scudo runtime.

It turns out it already is. On [Linux](https://github.com/apple/swift/blob/eae8d028ba469e11941dc6939b34fbb26c7d5fea/lib/Driver/UnixToolChains.cpp#L291-L301) the driver will simply pass the sanitizer option through to clang, which drives the actual linking process. This results in the sanitising being applied. Again, I've verified this in live code.

> This is missing tests.

Fantastic. Can you point me to places in the code that do similar testing? I wasn't able to find them, and that made it very hard to write more.

danliew-apple commented:
@Lukasa Here are some examples:

* [Here's an example driver test that checks that the compiler refuses certain flags](https://github.com/apple/swift/blob/master/test/Driver/sanitize_recover.swift). The test is also checking that the driver passes an option to the Swift frontend. That isn't relevant in this case so you can ignore that.
* [This test](https://github.com/apple/swift/blob/master/test/Driver/sanitizers.swift) does a bunch of things but one the things it does is check that a sanitizer runtime library is being passed to the linker.
* [Here's an example of a test that builds and executes code](https://github.com/apple/swift/blob/master/test/Sanitizers/asan_recover.swift)

Note that `llvm-lit` has a notion of features and you can use this to make a test run on only certain platforms. My suggestion is that

* Tests for driver options should probably execute on all platforms. In these tests you'd explicit specify the target so that we can verify that the driver allows Scudo on Linux but rejects it for Windows on macOS.
* Tests that are only going to work on Linux should be annotated with `REQUIRES: OS=linux-gnu`.

Lukasa commented:
Thanks @danliew-apple, I’ll write some tests to cover this and ping for re-review when I’ve done so.

danliew-apple commented:
> I don't believe this is true. There are separate Sanitizer whitelists for Darwin and Windows which explicitly whitelist the sanitisers they support. I have observed that with this toolchain built on Darwin the Scudo flag is rejected: I believe this is true for Windows as well, though @compnerd could further clarify.

The code you link to is related to linking. The **driver** should fail compilation much earlier than that.  If someone passes `-sanitizer=scudo` option for any more of the driver and the platform doesn't support it then it should fail.

> This patch doesn't change anything about the compilation process. AFAIK Scudo is a replacement allocator library so I would expect that the linker would need to be told to link against the scudo runtime.
>> It turns out it already is. On Linux the driver will simply pass the sanitizer option through to clang, which drives the actual linking process. This results in the sanitising being applied. Again, I've verified this in live code.

Interesting. I would not have expected that given that the Swift compiler shouldn't know what to do with this new enum you've added. You'll need to show me working tests to convince me.

danliew-apple commented:
@Lukasa Just to be clear I'm only giving you a review the code as I see it. I don't work on the swift compiler very frequently so I think you will need approval from others to land this. Given that this is very untested I suspect we might want to label it as experimental.

Lukasa commented:
@danliew-apple Totally understood, and I’m very appreciative of your time. I’m confident the patch-as-written works in the straightforward cases because I tested it live, but I also totally agree that without the tests it is hard for anyone else to validate my confidence! So I absolutely appreciate the feedback. I’ll work on adding those tests, and we can worry about wider approval once I’ve dealt with this feedback.

danliew-apple commented:
@airspeedswift @mikeash Does adopting the hardened Scudo on Linux seem a reasonable thing to support?

danliew-apple commented:
@Lukasa I just noticed that the Scudo documentation says.. "Compiling with Scudo will also enforce PIE for the output binary". That's the behaviour for Clang. Do you know why they do this? If there's a good reason for this we might need to do the same thing for Swift. I don't know if Swift has an equivalent PIE mode.

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

mikeash commented:
@danliew-apple I'm not too familiar with it, but reading through the link you provided, this seems like a great thing to support. Especially since the hard work is in LLVM and we apparently just need to support the option and pass it along. @weissi Any additional thoughts?

weissi commented:
@mikeash / @Lukasa this looks good to me, IMHO definitely something we want :)

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please test

Lukasa commented:
Solid work Jenkins, thanks for that.

Lukasa commented:
@danliew-apple I believe I have covered all the bases with testing here. I have validated that the driver correctly rejects the Scudo "sanitizer" on all non-Linux platforms, and that Scudo is successfully enabled on Linux and works (by looking for a Scudo-specific error message on a double-free). Hopefully the test coverage that exists here will be sufficient to convince folks that the functionality does work as intended.

Lukasa commented:
The OS X platform failure seems unrelated to this change, everything else is shiny and happy.

jckarter reviewed (approved):
LGTM, shouldn't affect anyone who isn't explicitly reaching for this flag.

yln commented on lib/Option/SanitizerOptions.cpp:
```diff
@@ -182,6 +182,36 @@ OptionSet<SanitizerKind> swift::parseSanitizerArgValues(
+    if (forbiddenOptions) {
+      SanitizerKind forbidden;
+
+      if (forbiddenOptions & SanitizerKind::Address) {
```
Note: this if-else needs to be extended whenever the SanitizerKind enum gets extended, but I don't see a better way of doing this with the current API of `OptionSet`.

Lukasa replied:
Yup. `OptionSet` absolutely could answer this question, but I don’t know how much value we get from ramming that new single-use API into this patch.

yln replied:
Not requesting anything additional for this PR, just wanted to point it out; maybe someone has an idea.

Lukasa replied:
Ah sorry, I understood you weren’t asking for new API, I was just adding clarification on why I didn’t add it as part of this patch.

danliew-apple commented on test/Driver/sanitize_scudo.swift:
```diff
@@ -0,0 +1,38 @@
+ */
+// RUN: %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -emit-library -sanitize=scudo -target x86_64-unknown-linux-gnu %s 2>&1 | %FileCheck --implicit-check-not=SCUDO_LINUX %s
+
+// SCUDO_LINUX: swift
```
Are these on the same line? If so this should probably be `SCUDO_LINUX-SAME:`.

danliew-apple commented on test/Driver/sanitize_scudo.swift:
```diff
@@ -0,0 +1,38 @@
+// RUN: %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -emit-library -sanitize=scudo -target x86_64-unknown-linux-gnu %s 2>&1 | %FileCheck --implicit-check-not=SCUDO_LINUX %s
+
+// SCUDO_LINUX: swift
+// SCUDO_LINUX: -pie
```
I'd still like an answer on this. As I understand it using `-pie` requires that the objects files be position independent code but I have idea if the Swift compiler actually guarantees this.

jckarter replied:
Swift should always use pie.

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
Hmm, the CI is very sad.

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
Thanks for the review @danliew-apple, I think I've addressed your feedback.

compnerd commented:
Hmm, do we not want the runtime to have a scudo variant as well if we are going to support this?  Having two memory domains is usually a good recipe for problems (what happens if someone does a `UnsafeMemoryPointer.deallocate` from Swift on something allocated from C with scudo intercepting the allocation?)

Lukasa commented:
Nope, there is one and only one allocator in a program at any one time. If scudo is on, it is on for everything, because it is linked into the main binary. In some senses this is a strength over ASan: as it operates at the allocator level there is no reason or need to provide customised runtimes or sanitizer hooks.

I encourage you to try it out: write a Swift package that allocates in C, frees in C, and is then deallocated again in Swift (for a double free). Scudo will catch it.

Lukasa commented:
Hey @danliew-apple, do you have any chance of looking at this this week?

Lukasa commented:
A gentle ping to @danliew-apple.

danliew-apple commented on test/Driver/sanitize_scudo.swift:
```diff
@@ -0,0 +1,38 @@
+// RUN: not %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=scudo -target arm64-apple-tvos9.0 %s 2>&1 | %FileCheck -check-prefix=SCUDO_tvOS %s
+// RUN: not %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=scudo -target armv7k-apple-watchos2.0 %s 2>&1 | %FileCheck -check-prefix=SCUDO_watchOS %s
+// RUN: not %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=scudo -target i386-apple-watchos2.0 %s 2>&1 | %FileCheck -check-prefix=SCUDO_watchOS_SIM %s
+// RUN: not %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=scudo -target x86-apple-macosx10.9 %s 2>&1 | %FileCheck -check-prefix=SCUDO_OSX_32 %s
```
This target triple is incorrect. It should be `i386-apple-macosx10.9`.

Lukasa replied:
Really? TSAN uses this one as well, we should consider fixing that up.

danliew-apple replied:
Yes. The Swift compiler rejects the `x86-apple-macosx10.9` target triple. And yes if you copy and pasted this from a TSan test we should fix that too but in a separate patch.

danliew-apple commented on test/Driver/sanitize_scudo.swift:
```diff
@@ -0,0 +1,38 @@
+/*
+ * Make sure we don't accidentally add the sanitizer library path when building libraries or modules
+ */
+// RUN: %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -emit-library -sanitize=scudo -target x86_64-unknown-linux-gnu %s 2>&1 | %FileCheck --implicit-check-not=SCUDO_LINUX %s
```
This `--implicit-check-not=SCUDO_LINUX` line doesn't look right. That's adding checks that we don't see `SCUDO_LINUX` in the compiler's output which doesn't make sense.

I presume you meant `-check-prefix=SCUDO_LINUX --implicit-check-not=<some pattern>`?

Where `<some pattern>` is whatever you don't what to appear in the compiler's output.

Lukasa replied:
I think this is right: when building only libraries, we don't want to add the sanitiser because it should only be added when building the main binary. This is the same as the requirement on all the other sanitisers with runtime libraries: only the main binary actually invokes the linker to link the sanitiser runtime.

danliew-apple commented on test/Driver/sanitize_scudo.swift:
```diff
@@ -0,0 +1,38 @@
+// SCUDO_LINUX: bin/clang
+// SCUDO_LINUX-SAME: -pie
+// SCUDO_LINUX-SAME: -fsanitize=scudo
+// SCUDO_OSX_32: unsupported option '-sanitize=scudo' for target 'x86-apple-macosx10.9'
```
You need to update the target triple here too.

danliew-apple commented:
@Lukasa Sorry for the delay in reviewing this.

Lukasa commented:
Thanks @danliew-apple, I've responded to one or two of your points where I think the patch is currently correct, and have addressed your others, including updating the commit message. Do you mind taking another look?

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
@swift-ci please test

Lukasa commented:
@swift-ci please smoke test

Lukasa commented:
Thanks for that clarification @danliew-apple, that helped a lot! I have rewritten that test and also updated the commit message, please take another look.

Lukasa commented:
@swift-ci please test

danliew-apple commented:
@swift-ci please test

danliew-apple commented:
@Lukasa Sorry for being slow. I've approved the changes. Thanks for being patient.

Lukasa commented:
No need to apologise, I understand completely that we’re all busy! Thanks so much for your time on this.

danliew-apple commented:
@swift-ci please test

compnerd commented:
I believe that this has caused a regression on Windows: https://ci-external.swift.org/job/oss-swift-windows-x86_64/2607/

Lukasa commented:
Good catch @compnerd, that fix looks right. Sorry about that!
